Code/Engine/RendererCore/Material/Implementation/MaterialManager.cpp

```diff
@@ -0,0 +1,686 @@
+    change.m_NewId = pMaterial->m_MaterialId;
+    s_MaterialShaderChangedEvent.Broadcast(change);
+  }
+  else
+  {
+    // Material reloaded with same shader. Nothing to do except marking material dirty below.
+  }
+  m_ModifiedMaterials.Insert(pMaterial->GetResourceHandle());
+}
+
+void ezMaterialManager::UnregisterMaterial(ezMaterialResource* pMaterial)
+{
+  EZ_LOCK(m_MaterialShaderMutex);
+  EZ_LOCK(m_ExtractionMutex);
+
+  MaterialShaderConstants& msc = GetShaderConstants(pMaterial->m_hShader);
+  msc.RemoveMaterial(pMaterial->m_MaterialId);
+  m_RemovedMaterials.PushBack(pMaterial);
+}
+
+void ezMaterialManager::ExtractMaterial(ezMaterialResource* pMaterial, ezMaterialManager::ExtractedMaterial& extractedMaterial)
+{
+  EZ_PROFILE_SCOPE("ExtractMaterial");
+  extractedMaterial.m_hMaterial = pMaterial->GetResourceHandle();
+  extractedMaterial.m_hShader = pMaterial->m_mDesc.m_hShader;
+  extractedMaterial.m_MaterialId = pMaterial->m_MaterialId;
+  extractedMaterial.m_DirtyFlags = pMaterial->m_DirtyFlags;
+  for (ezMaterialResource::DirtyFlags::Enum flag : pMaterial->m_DirtyFlags)
+  {
+    switch (flag)
+    {
+      case ezMaterialResource::DirtyFlags::Parameter:
+        extractedMaterial.m_Parameters = pMaterial->m_mDesc.m_Parameters;
+        break;
+      case ezMaterialResource::DirtyFlags::Texture2D:
+        extractedMaterial.m_Texture2DBindings = pMaterial->m_mDesc.m_Texture2DBindings;
+        break;
+      case ezMaterialResource::DirtyFlags::TextureCube:
+        extractedMaterial.m_TextureCubeBindings = pMaterial->m_mDesc.m_TextureCubeBindings;
+        break;
+      case ezMaterialResource::DirtyFlags::PermutationVar:
+        extractedMaterial.m_PermutationVars = pMaterial->m_mDesc.m_PermutationVars;
+        break;
+      default:
+        break;
+    }
+  }
+  pMaterial->m_DirtyFlags.Clear();
+}
+
+void ezMaterialManager::ApplyMaterialChanges()
+{
+  if (m_pPendingChanges == nullptr)
+  {
+    // If no pending changes are present, we will start extracting here.
+    // This is usually the case for applications that don't use extraction like tests and basic sample apps.
+    ExtractMaterialUpdates();
+  }
+  EZ_PROFILE_SCOPE("ApplyMaterialChanges");
+  // Remove dead pointers first
+  for (const void* pRemovedMaterial : m_pPendingChanges->m_RemovedMaterials)
+  {
+    m_Materials.Remove(pRemovedMaterial);
+  }
+
+  // Execute updates and additions
+  for (const auto& extractedMaterial : m_pPendingChanges->m_AddedOrModifiedMaterials)
+  {
+    ezResourceLock<ezMaterialResource> pMaterial(extractedMaterial.m_hMaterial, ezResourceAcquireMode::PointerOnly);
+    auto matIt = m_Materials.FindOrAdd(pMaterial.GetPointer());
+    MaterialData& md = matIt.Value();
+    for (ezMaterialResource::DirtyFlags::Enum flag : extractedMaterial.m_DirtyFlags)
+    {
+      switch (flag)
+      {
+        case ezMaterialResource::DirtyFlags::Parameter:
+          md.m_Parameters = extractedMaterial.m_Parameters;
```

Comment:
Could use move here instead of copy since the extracted material will be dead anyways afterwards, right?

Reply:
Not trivially, because these are using the frame allocator but you are right, these can be moved if I fix the allocator to use for these.

Reply:
Fixed!